### PR TITLE
wip - fix dict word predictor

### PR DESCRIPTION
**Pull request by kyleclo**
1. define a new whitespace tokenization predictor
2. add IDs to rows and pages in PDFPlumber
3. add `tokenizers` to setup
4. add new DictWordPredictor logic

### CODE REVIEW

**Review by comorado — APPROVED** [review]

File: src/mmda/predictors/heuristic_predictors/dictionary_word_predictor.py [hunk 1]
```diff
@@ -1,27 +1,76 @@
...
+            return text.strip().lower() in self.words
+
+    def strip_punct(self, text: str) -> str:
+        start = 0
+        while text[start] in self.punct and start < len(text):
+            start += 1
+        end = len(text) - 1
+        while text[end] in self.punct and end > 0:
+            end -= 1
+        return text[start: end + 1]
```
Comment (comorado) [inline src/mmda/predictors/heuristic_predictors/dictionary_word_predictor.py:62]:
```suggestion
        assert start >= end
        return text[start: end + 1]
```

File: src/mmda/predictors/heuristic_predictors/dictionary_word_predictor.py [hunk 2]
```diff
@@ -80,113 +124,301 @@ def predict(self, document: Document) -> List[SpanGroup]:
...
+            token_id_to_word_id=token_id_to_word_id,
+            word_id_to_text=word_id_to_text
+        )
+
+        # 5) cleanup
+        document.remove(field_name='_ws_tokens')
+        return words
 
-                words.append(self._copy_token_with_text(token))
+    def _precompute_token_features(self, document: Document) -> Tuple:
```
Comment (comorado) [inline src/mmda/predictors/heuristic_predictors/dictionary_word_predictor.py:165]: Any chance this function can be broken up? It is hard to follow the logic

**Review by soldni — APPROVED** [review]
Looks good! Bump mmda version?

File: src/mmda/predictors/heuristic_predictors/dictionary_word_predictor.py [hunk 3]
```diff
@@ -80,113 +124,301 @@ def predict(self, document: Document) -> List[SpanGroup]:
...
+
+        # beginning/end of row w/ hyphen
+        # TODO: add pages too
+        row_end_with_hyphen_token_ids = set()
+        row_start_after_hyphen_token_ids = set()
+        max_row_end_token_id_to_min_row_start_token_id = {}
+        for i in range(0, len(document.tokens) - 1):
+            current = document.tokens[i]
+            next = document.tokens[i + 1]
+            is_transition = current.rows[0].id != next.rows[0].id
```
Comment (geli-gel) [inline src/mmda/predictors/heuristic_predictors/dictionary_word_predictor.py:206]: do we have to update pdfplumber for this? does it not work if we say like `current.rows[0] != next.rows[0]`? just so we don't have to update pdfplumber as well

File: src/mmda/predictors/heuristic_predictors/dictionary_word_predictor.py (hunk 1 above)
Comment (kyleclo) [reply src/mmda/predictors/heuristic_predictors/dictionary_word_predictor.py:62]: why is this necessary?

File: src/mmda/predictors/heuristic_predictors/dictionary_word_predictor.py (hunk 2 above)
Comment (kyleclo) [reply src/mmda/predictors/heuristic_predictors/dictionary_word_predictor.py]: i've split into 2 - whitespace tokenization & token-level features

File: src/mmda/predictors/heuristic_predictors/dictionary_word_predictor.py (hunk 3 above)
Comment (kyleclo) [reply src/mmda/predictors/heuristic_predictors/dictionary_word_predictor.py:220]: yea, unfortunately this would make things much easier to write

File: src/mmda/predictors/heuristic_predictors/dictionary_word_predictor.py [hunk 4]
```diff
@@ -1,27 +1,76 @@
...
+import os
+from typing import Optional, Set, List, Tuple, Iterable, Dict
 
+from collections import defaultdict
+from mmda.parsers import PDFPlumberParser
 from mmda.predictors.base_predictors.base_predictor import BasePredictor
-from mmda.types.metadata import Metadata
-from mmda.types.annotation import Annotation, Span, SpanGroup
-from mmda.types.document import Document
+from mmda.predictors import WhitespacePredictor
```
Comment (geli-gel) [inline src/mmda/predictors/heuristic_predictors/dictionary_word_predictor.py:34]: I think this needs to be `from mmda.predictors.heuristic_predictors import WhitespacePredictor`, right?

File: src/mmda/predictors/heuristic_predictors/dictionary_word_predictor.py (hunk 1 above)
Comment (comorado) [reply src/mmda/predictors/heuristic_predictors/dictionary_word_predictor.py:62]: It is not necessary disregard, I thought maybe it is a good check

**Review by comorado — APPROVED** [review]

**Comment by kyleclo** [comment]
waiting @geli-gel approval to make sure doesnt break anything in orchestration code
